src/lava/lib/optimization/utils/solver_tuner.py

```diff
@@ -26,37 +27,41 @@ def __init__(self, params_grid: ty.Dict):
     def tune(self,
              solver,
              solver_parameters: ty.Dict,
-             stopping_condition: ty.Callable[[float, int], bool] = None):
-        """Find and set solving hyperparameters for an OptimizationSolver"""
+             stopping_condition: ty.Callable[[float, int], bool] = None,
+             seed: int = 0):
+        """
+        Perform random search to optimize solver hyper-parameters.
+        """
         # TODO : Check that hyperparameters are arguments for solver
         best_hyperparameters, best_cost, best_step_to_sol = self._random_search(
             solver,
             solver_parameters,
-            stopping_condition)
+            stopping_condition,
+            seed=seed)
         succeeded = stopping_condition(best_cost, best_step_to_sol)
         return best_hyperparameters, succeeded
 
     def _random_search(self,
                        solver,
                        solver_parameters: ty.Dict,
                        stopping_condition: ty.Callable[
-                           [float, int], bool] = None
+                           [float, int], bool] = None,
+                       seed: int = 0
                        ) -> ty.Union[ty.NoReturn, ty.Dict]:
 
         best_hyperparameters = None
         best_cost = float("inf")
         best_step_to_sol = float("inf")
-        problem = solver.problem
+
         params_names = self._params_grid.keys()
         params_grid = list(
             it.product(*map(lambda k: self._params_grid[k], params_names)))
-        random.shuffle(params_grid)
+        random.Random(seed).shuffle(params_grid)
 
         for params in params_grid:
             hyperparameters = dict(zip(params_names, params))
             solver_parameters["hyperparameters"] = hyperparameters
-            solution = solver.solve(**solver_parameters)
-            # TODO : Implement logic for CSP problems
+            solver.solve(**solver_parameters)
             cost = solver.last_run_report["cost"]
             step_to_sol = solver.last_run_report["steps_to_solution"]
             if cost is not None and cost <= best_cost and (

```

Comment:
I was thinking about this metric. Still, it feels weird to consider "best" according to our arbitrary metric (cost AND time must be minimum), instead of asking for a user input.

What's about this option:
The user defines a metric in the following way:

`def stop(cost, time) -> ty.Tuple[float, bool]:`

The first return value provides the 1D metric used to define what "best solution" means. The second return value determines if the stopping criteria has been fullfilled. For example, if I want to find the optimal solution, my function look look like

`def stop(cost, time):
    cost_opt = 10
    return cost - cost_opt, cost == cost_opt
`

Then the SovlerTuner will provide you with the hyperparameter set that minimizes the first output (cost-cost_opt), and stops if the second condition has been fulfilled.

@GaboFGuerra , what do you think?

Reply:
i agree on the "weird metric" part. I would probably go for a *fitness function* as you are suggesting, and then a target value for the fitness function to stop the execution.

```python

    def fitness(cost, steps): 
      return cost`
    
    # SolverTuner API
    def tune(self,
             solver,
             solver_parameters: ty.Dict,
             fitness: ty.Callable[[float, int], float],
             target_fitness: float = None,
             seed: int = 0):
```